### ISSUE TITLE
Extra blit required on XR devices with one framebuffer per eye?

### ISSUE THREAD

**Issue by asajeffrey** [immersive-web/webxr#968]
The `XRWebGLLayer` interface https://immersive-web.github.io/webxr/#xrwebgllayer-interface provides a single framebuffer, with one viewport for each eye. When implementing this API for devices (e.g HoloLens or MagicLeap) which have a framebuffer for each eye, this means there's an extra blit where the WebXR framebuffer is copied into each device framebuffer.

Are there plans for an API which can be implemented on such devices without an extra blit? For example by having a method `getFramebuffer(view)` similar to `getViewport(view)` rather than a single `framebuffer` property?

**Comment by cabanier** [comment]
The layers proposal will give you direct access to texture that you can draw into.
See https://github.com/immersive-web/layers/, specifically:
`  Promise<XRProjectionLayer> requestProjectionLayer(GLenum textureTarget, XRProjectionLayerInit init);`
and
`  XRWebGLSubImage? getViewSubImage(XRLayer layer, XRView view)`

**Comment by asajeffrey** [comment]
OK, so with that API a user would use `glLayerFactory.requestProjectionLayer(gl.TEXTURE_2D_ARRAY)` to request l layer, then `getViewSubImage(layer, view)` to get a subimage, then `subimage.colorTexture` to get the texture array?

Are there restrictions similar to those for opaque framebuffers as to how those textures can be used? In particular, is there a requirement that if they're used outside of the XR rAF in which they were accessed, then all bets are off?

**Comment by cabanier** [comment]
> OK, so with that API a user would use `glLayerFactory.requestProjectionLayer(gl.TEXTURE_2D_ARRAY)` to request l layer, then `getViewSubImage(layer, view)` to get a subimage, then `subimage.colorTexture` to get the texture array?

Yes

> Are there restrictions similar to those for opaque framebuffers as to how those textures can be used? In particular, is there a requirement that if they're used outside of the XR rAF in which they were accessed, then all bets are off?

Yes, these buffers will only be valid during the Raf callback that they are received.

**Comment by toji** [comment]
The intent is for them to effectively be used like a swap chain, with a new texture provided every frame and the old ones becoming inaccessible as they're consumed by the compositor.  Exact lifetime of the texture object still needs to be determined. Otherwise they should have far less restrictions than the opaque framebuffer equivalents, in that you'll be able to use them like any other texture during their active lifetime.

**Comment by asajeffrey** [comment]
OK, This does require WebGL2, which is a pity. Are there plans for a similar WebGL1 API?

**Comment by toji** [comment]
You'll be able to use WebGL 1 contexts with the layers API, but you won't be able to produce an Array texture for obvious reasons. At the moment in WebGL 1 I was assuming that you'd get a singe texture and write each eye to a different viewport, like the opaque framebuffer (but with no antialiasing). Now that I'm thinking about it, though, if there was an advantage to returning one texture per view for WebGL 1 contexts that may be worth doing. I think that might allow blits to be eliminated with OpenXR, but possibly not on the native Oculus API or Windows MR? Not sure about Magic Leap or other platforms.

**Comment by asajeffrey** [comment]
Yes, if the spec allowed for `glLayerFactory.requestProjectionLayer(gl.TEXTURE_2D)` to return a layer with different textures for each eye, then this would eliminate a blit on openxr and magicleap. That would be great!

**Comment by toji** [comment]
Good to know, we'll consider that! I should make it clear that I expect very few people to use a projection layer with WebGL 1, though, because it won't have an antialiasing mechanism. `XRWebGLLayer` will probably still be the dominant layer type with that API.

**Comment by cabanier** [comment]
> At the moment in WebGL 1 I was assuming that you'd get a singe texture and write each eye to a different viewport, like the opaque framebuffer (but with no antialiasing). 

The API still allows WebGL1.
Calling `getViewSubImage` for each eye might return the same colorTexture with a different viewpoer or 2 different textures.

> Now that I'm thinking about it, though, if there was an advantage to returning one texture per view for WebGL 1 contexts that may be worth doing. I think that might allow blits to be eliminated with OpenXR, but possibly not on the native Oculus API or Windows MR? Not sure about Magic Leap or other platforms.

Oculus doesn't really care. It supports texture array, an array of textures or a side by side texture.
I thought Magic Leap only supported texture arrays. Is that correct, @raviramachandra

**Comment by toji** [comment]
> Calling getViewSubImage for each eye might return the same colorTexture with a different viewpoer or 2 different textures.

That's a really good point! We don't have to return actual arrays of textures, just a unique one per view. I kinda feel like we'd want to specify one or the other, though, since that's the kind of behavior that causes content to break across devices. (If I'm on platform A and it gives me the same texture each view I might be tempted to bind it once at the top of the render loop and then never check the others, which would break on platform B)

**Comment by asajeffrey** [comment]
The example at the beginning of https://github.com/immersive-web/layers/#rendering probably needs rewritten then. It's not obvious what `glLayerFactory.getSubImage(layer)` (the version without a `view`) is for.

**Comment by cabanier** [comment]
> We don't have to return actual arrays of textures, just a unique one per view. I kinda feel like we'd want to specify one or the other, though, since that's the kind of behavior that causes content to break across devices.

I wanted to leave that up to the UA in case one implementation is faster than the other. (ie single screen vs a screen per eye) but I don't feel very strongly

**Comment by raviramachandra** [comment]
> I thought Magic Leap only supported texture arrays. Is that correct, @raviramachandra

That is correct.
